diff --git a/bpf/fillers.h b/bpf/fillers.h
index 2779a2a..b34a28a 100644
--- a/bpf/fillers.h
+++ b/bpf/fillers.h
@@ -2003,10 +2003,6 @@ FILLER(proc_startupdate_3, true)
 		 * loginuid
 		 */
 		/* TODO: implement user namespace support */
-		loginuid = _READ(task->loginuid);
-		res = bpf_val_to_ring_type(data, loginuid.val, PT_INT32);
-		if (res != PPM_SUCCESS)
-			return res;
 	}
 
 	return res;
